core/src/consensus.rs

```diff
@@ -454,6 +455,40 @@ impl Tower {
         self.last_vote_tx_blockhash
     }
 
+    pub fn refresh_last_vote_timestamp(
+        &mut self,
+        heaviest_bank_on_same_fork: Slot,
+        slot_duration: Duration,
+    ) {
+        let last_vote_timestamp = self.last_vote.timestamp();
+        if let Some(last_vote_timestamp) = last_vote_timestamp {
+            // To avoid a refreshed vote tx getting caught in deduplication filters,
+            // we need to update timestamp. Increment by smallest amount to avoid skewing
+            // the Timestamp Oracle.
+            self.last_vote
+                .set_timestamp(Some(last_vote_timestamp.saturating_add(1)));
+        } else if let Some(last_voted_slot) = self.last_vote.last_voted_slot() {
+            // If the previous vote did not send a timestamp due to clock error, try our
+            // best to estimate the correct timestamp for the Timestamp Oracle.
```

Comment:
ick is this because this `maybe_timestamp` might return `None` here: https://github.com/ilya-bobyr/solana/blob/a872c0a373015a5427bdce238b32d66783595378/core/src/consensus.rs#L538

Shouldn't the ` if current_slot > self.last_timestamp.slot` always be true since votes are always increasing? Seems like this case we can just submit the current timestamp since it should never be true

Reply:
hmmm how often is `Utc::now().timestamp()` going to go backwards? Feel like we could just panic in that case or retry until we get a proper timestamp. Estimating the slot time here seems kind of hacky.

Reply:
🤷. also if it was messed up for the original vote it could also be messed up for the estimation here. 
do we want to kill everything over a missed timestamp though?

Reply:
> how often is Utc::now().timestamp() going to go backwards?

whenever it wants. underlying time is sampled from `std::time::SystemTime` (`clock_gettime(CLOCK_REALTIME)` on linux), which is not monotonic. `std::time::Instant` uses `clock_gettime(CLOCK_MONOTONIC)`, but i don't think there's any way to get a timestamp from one.

i wonder if we'd be better off looking up the txs' rbh in status cache and comparing the slots to tiebreak. abusing the wallclock oracle is starting to feel too dirty

Reply:
Or is there a bit in the VoteState we can repurpose as a counter?

Reply:
> i wonder if we'd be better off looking up the txs' rbh in status cache and comparing the slots to tiebreak

We thought about doing this as well, but it's a pain because 

1. Blockhashes can expire
2. Each bank's blockhash queue only has the blocks on its fork, so if the `nth` refresh and `n+1th` refresh use blockhashes from different forks, you'll have to do a search

Reply:
> > i wonder if we'd be better off looking up the txs' rbh in status cache and comparing the slots to tiebreak
> 
> We thought about doing this as well, but it's a pain because
>     1. Blockhashes can expire
>     2. Leader may not have played the block yet, so doesn't have the blockhash
>     3. Each bank's blockhash queue only has the blocks on its fork, so if the `nth` 
refresh and `n+1th` refresh use blockhashes from different forks, you'll have to do a search

don't these conditions all imply the same thing? that is, "this vote transaction will not land on this fork as of my view atm." which means there's no reason to cache it. in that case i think status cache will return `None` and since `None` is strictly less than `Some(_)`, we get the priority that we want

Reply:
Think that's fine for 1.  However for 2, since we start ingesting votes a little earlier than our leader slot this might result in some votes being dropped when they would have landed.
3 would require some work to determine which blockhash is older but doable. 

Ideally we would add another field or change timestamp to enum, but waiting another year for `VoteStateUpdateV2` hurts. I think abusing timestamp for now is our best bet.

Reply:
How important is this check to ensure that timestamps don't go backwards? I'm guessing it's important to ensure that the average clock doesn't go backwards?

```
fn maybe_timestamp(&mut self, current_slot: Slot) -> Option<UnixTimestamp> {
        if current_slot > self.last_timestamp.slot
            || self.last_timestamp.slot == 0 && current_slot == self.last_timestamp.slot
        {
            let timestamp = Utc::now().timestamp();
            if timestamp >= self.last_timestamp.timestamp {
                self.last_timestamp = BlockTimestamp {
                    slot: current_slot,
                    timestamp,
                };
                return Some(timestamp);
            }
        }
        None
    }
```

Because it seems this new `best_estimate`  doesn't ensure that it's greater than the `self.last_timestamp` either. In which case maybe it's better to handle it at the `maybe_timestamp` layer by just setting it to something like ``self.last_timestamp + 1`

Reply:
timestamp oracle uses a stake weighted median so an erroneous `self.last_timestamp + 1` once in a blue moon shouldn't affect it.

Reply:
@CriesofCarrots any thoughts about abusing the timestamp field like this?

Reply:
It's not my first choice fix, but I understand the desire to do this without changing the Vote struct (ie. feature gate). Based on the rates of vote refreshes that Ashwin described to me over DM, I don't think that adding a second here or there (or even 33% of stake) is likely to have a large impact on the timestamp oracle output.

I think we do need to ensure `last_timestamp` is always increasing. The vote program will reject a vote that has a timestamp going backward in time.

Reply:
> It's not my first choice fix, but I understand the desire to do this without changing the Vote struct (ie. feature gate). Based on the rates of vote refreshes that Ashwin described to me over DM, I don't think that adding a second here or there (or even 33% of stake) is likely to have a large impact on the timestamp oracle output.

right. total gossip refreshes shouldn't be more than `floor(slot_hashes::MAX_ENTRIES / MAX_PROCESSING_AGE)` today that's `floor(512 / 150)` -> `3`. i do not believe however that we're limited to 33% of stake refreshing.

> I think we do need to ensure `last_timestamp` is always increasing. The vote program will reject a vote that has a timestamp going backward in time.

yeah this is a concern of mine as well. i think we'll need to not vote on slots with timestamps earlier than our last refresh timestamp or else risk a compounding effect that _does_ skew the oracle. does that sound right?

Reply:
> yeah this is a concern of mine as well. i think we'll need to not vote on slots with timestamps earlier than our last refresh timestamp or else risk a compounding effect that does skew the oracle. does that sound right?

the existence of a refreshed vote implies that we haven't voted past our original vote. Using "+1" here ensures that that timestamp is increasing. Additionally if we ever see a backwards timestamp in `maybe_set_timestamp` we can also use a "+1" from our last vote to ensure timestamp is increasing in the non-refresh case.

Reply:
> the existence of a refreshed vote implies that we haven't voted past our original vote. Using "+1" here ensures that that timestamp is increasing.

yeah i'm interested in the votes immediately after a refresh successfully lands

> Additionally if we ever see a backwards timestamp in `maybe_set_timestamp` we can also use a "+1" from our last vote to ensure timestamp is increasing in the non-refresh case.

this is the "compounding effect" i was referring to

Reply:
I'm worried we're over fixating on this compounding effect here.

What is the SLA for the timestamp oracle? From what i've pieced together it is used to approximate when a block has been frozen by the cluster for regulatory authorities so it should ideally pretty accurate especially in "interesting" times aka consensus problems? 

If any validators have to refresh, or any validator has to switch then we've already screwed the oracle. Timestamps are generated when the vote tx is generated - they will be off by 10-15 secs for the vote following a refresh, or any # of time depending on how close the switch slot is to the tip of the new fork if we're waiting for switch proof.

Depending on the importance of this it might be better to repurpose the timestamp entirely and use the bank frozen time from replay?

Reply:
Bank-frozen time actually makes a lot of sense for the timestamp oracle.
Would you still want to increment by 1sec for the vote-refresh distinction?

Reply:
Yeah, my idea is
* Populate timestamp when freezing bank
* If UTC::now fails, use the best_estimate calc. More likely to be correct because there is no vote delay, and it's an inverse of the calc used in the stake weighted mean anyway. 
* If we're worried about the compounding effect, denote which timestamps have been estimated and which have been observed. run best_estimate from the latest bank that has been observed.
* For refresh, during vote generation add a +1

Reply:
Unless I'm missing something, I would expect the estimate off an estimate to be the same as an estimate off the original latest observed bank.

But anyway, I think this approach sounds okay, with the caveat that we probably still need to ensure that the timestamp doesn't go backward immediately after a refresh, whether the new vote timestamp is generated by UTC::now or best_estimate, so that those votes do not fail to land.

Reply:
I agree, seems way too much to backport. For now i've stuck with the `+1` idea, and updated `last_timestamp` such that new vote timestamps cannot go backwards. Vote ts will be increasing or `None`. 

for the future: https://github.com/solana-labs/solana/issues/32135 to track how to fix this properly.